### PR TITLE
Enable mypy-zope plugin

### PR DESCRIPTION
**Pull request by wsanchez**
Enable Mypy-Zope so that we can properly type check the use of Zope Interface classes and remove the various hacks we have in place to make Mypy happy without the plugin.

### CODE REVIEW

**Comment by wsanchez** [comment]
Getting an INTERNAL ERROR after just enabling it:

```console
GLOB sdist-make: /home/runner/work/klein/klein/setup.py
mypy create: /home/runner/work/klein/klein/.tox/mypy
mypy installdeps: mypy==0.812, mypy-zope==0.2.11, Twisted==20.3.0, attrs==20.3.0, Automat==20.2.0, characteristic==14.3.0, constantly==15.1.0, hyperlink==21.0.0, incremental==17.5.0, PyHamcrest==2.0.2, six==1.15.0, Tubes==0.2.0, Werkzeug==1.0.1, zope.interface==5.2.0
mypy inst: /home/runner/work/klein/klein/.tox/.tmp/package/1/klein-20.6.0.zip
mypy installed: attrs==20.3.0,Automat==20.2.0,characteristic==14.3.0,constantly==15.1.0,hyperlink==21.0.0,idna==3.1,incremental==17.5.0,klein @ file:///home/runner/work/klein/klein/.tox/.tmp/package/1/klein-20.6.0.zip,mypy==0.812,mypy-extensions==0.4.3,mypy-zope==0.2.11,PyHamcrest==2.0.2,six==1.15.0,Tubes==0.2.0,Twisted==20.3.0,typed-ast==1.4.2,typing-extensions==3.7.4.3,Werkzeug==1.0.1,zope.event==4.5.0,zope.interface==5.2.0,zope.schema==6.1.0
mypy run-test-pre: PYTHONHASHSEED='688791434'
mypy run-test: commands[0] | mypy --cache-dir=/home/runner/work/klein/klein/.tox/mypy_cache release.py setup.py src
src/klein/test/test_trial.py:33: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at https://github.com/python/mypy/issues
version: 0.812
src/klein/test/test_trial.py:33: : note: please use --show-traceback to print a traceback when reporting a bug
ERROR: InvocationError for command /home/runner/work/klein/klein/.tox/mypy/bin/mypy --cache-dir=/home/runner/work/klein/klein/.tox/mypy_cache release.py setup.py src (exited with code 2)
___________________________________ summary ____________________________________
ERROR:   mypy: commands failed
Error: Process completed with exit code 1.
```

File: src/klein/_session.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def procureSession(
             ]
             # Does it seem like this check is expensive? It sure is! Don't want
             # to do it? Turn on your dang HTTPS!
-            yield self._store.sentInsecurely(allPossibleSentTokens)
+            self._store.sentInsecurely(allPossibleSentTokens)
```
Comment (wsanchez) [inline src/klein/_session.py:126]: `sentInsecurely()` returns `None` and is not async, so mypy complains about this unnecessary `yield`.

File: src/klein/test/test_trial.py [hunk 2]
```diff
@@ -7,27 +7,26 @@
...
-        @ifmethod
-        def frob() -> None:
-            """
-            Frob the object.
-            """
+
+class TestCaseTests(TestCase):
+    """
+    Tests for L{TestCase}.
+    """
```
Comment (wsanchez) [inline src/klein/test/test_trial.py:29]: Moved to module scope to make `mypy-zope` happy.

File: src/klein/_isession.py [hunk 3]
```diff
@@ -321,19 +309,24 @@ def finalize() -> None:
         """
 
 
-class _IRequestLifecycle(Interface):
+class IRequestLifecycle(Interface):
     """
     Interface for adding hooks to the phases of a request's lifecycle.
     """
 
+    def addPrepareHook(
```
Comment (wsanchez) [inline src/klein/_isession.py:317]: Users of `IRequestLifecycle` are expecting this.

File: src/klein/storage/_memory.py [hunk 4]
```diff
@@ -55,7 +55,7 @@ class _MemoryAuthorizerFunction:
     Type shadow for function with the given attribute.
     """
 
-    __memoryAuthInterface__: IInterface = None
+    __memoryAuthInterface__: IInterface = None  # type: ignore[assignment]
```
Comment (wsanchez) [inline src/klein/storage/_memory.py:58]: Don't allow this to be set to `None` outside of init.

File: src/klein/_form.py [hunk 5]
```diff
@@ -458,6 +458,11 @@ class IProtoForm(Interface):
     Marker interface for L{ProtoForm}.
     """
 
+    def addField(field: Field) -> "FieldInjector":
```
Comment (wsanchez) [inline src/klein/_form.py:461]: Users of `IProtoForm` expect this.

**Review by adiroiban — APPROVED** [review]
Looks so much cleaner now :)

I think the interface updates might break in theory backward compatibility.
I have no idea how to implement bacward compatible interface changes.
I guess the only options is IProtoForm and then IProtoFormV2(IProtoForm)

I have no idea how many klein users are implementing their own forms.

Thanks!

File: src/klein/storage/_memory.py (hunk 4 above)
Comment (wsanchez) [inline src/klein/storage/_memory.py:58]: We don't want to encourage assigning `None` after initialization.

File: src/klein/test/test_request.py [hunk 6]
```diff
@@ -33,7 +33,10 @@ def requestFromBytes(data: bytes = b"") -> FrozenHTTPRequest:
 
     @classmethod
     def messageFromBytes(cls, data: bytes = b"") -> IHTTPMessage:
-        return cls.requestFromBytes(data)
+        # type note: This returns a FrozenHTTPRequest, which implements
+        # IHTTPRequest, which is a subclass of IHTTPMessage.
+        # Seems like a bug in mypy-zope.
+        return cls.requestFromBytes(data)  # type: ignore[return-value]
```
Comment (wsanchez) [inline src/klein/test/test_request.py:39]: File a bug on mypy-zope and update this comment with a link

File: src/klein/test/test_response.py [hunk 7]
```diff
@@ -30,7 +30,10 @@ def responseFromBytes(data: bytes = b"") -> FrozenHTTPResponse:
 
     @classmethod
     def messageFromBytes(cls, data: bytes = b"") -> IHTTPMessage:
-        return cls.responseFromBytes(data)
+        # type note: This returns a FrozenHTTPResponse, which implements
+        # IHTTPResponse, which is a subclass of IHTTPMessage.
+        # Seems like a bug in mypy-zope.
+        return cls.responseFromBytes(data)  # type: ignore[return-value]
```
Comment (wsanchez) [inline src/klein/test/test_response.py:36]: File a bug on mypy-zope and update this comment with a link

File: src/klein/storage/_memory.py (hunk 4 above)
Comment (wsanchez) [reply src/klein/storage/_memory.py:58]: I generally agree, though I think None is fine in this context and don't see there need to define an `IUndeclared`.

I added an assertion above the line were `__memoryAuthInterface__` is accessed to make sure it isn't still `None`.

**Comment by wsanchez** [comment]
> I think the interface updates might break in theory backward compatibility.
> I have no idea how to implement bacward compatible interface changes.
> I guess the only options is IProtoForm and then IProtoFormV2(IProtoForm)
> 
> I have no idea how many klein users are implementing their own forms.

There are still some mypy errors here, as there are a lot of these "marker interfaces" in play, and I don't understand why we have interfaces with zero defined API in them and then there's this injection magic, etc.

I don't use (or really understand) this API at all, so… shrug… it seems like the API I added to `IProtoForm`, for example, probably should really be there, as it's expected by the internal machinery, so I suspect that if you defined you own, that it would break at runtime when these attributes are accessed, if not due to the interface being incomplete according to Zope.  But again, I don't know what this really does.

@glyph, I think this is your creation, so it would be nice if you had a few cycles to clean up the last bits here…

**Comment by glyph** [comment]
> @glyph, I think this is your creation, so it would be nice if you had a few cycles to clean up the last bits here…

I'm (supposedly) taking the day off but filtering this kind of email is a project that is hard to bother with for less than a week's worth of vacation so I guess I might as well reply :).

As far as I can tell the interface changes that Adi raised concerns about are fine, because:

1. we weren't publishing any type stubs before, and
2. none of the code in question was actually *executing* at runtime.

I think I can find a couple of cycles next week to address the stuff in the implementation code but most of the stuff in test code is boring enough that I'll `cast` or `type: ignore` it away.

**Comment by glyph** [comment]
> it seems like the API I added to `IProtoForm`, for example, probably should really be there, as it's expected by the internal machinery, so I suspect that if you defined you own, that it would break at runtime when these attributes are accessed, if not due to the interface being incomplete according to Zope.

I agree with this, GVOS applies

**Comment by glyph** [comment]
> I don't understand why we have interfaces with zero defined API in them

*some* of them, as you observed, are just wrong, because mypy was previously providing no assistance and `Any`-ing us into apparent safety when it wasn't there.

However, some of them are to satisfy some mechanical requirements because they require more from the technology than it can currently provide.  For example: `IPrototoForm` is intentionally `Any`-ish because it's a dict with certain properties which are impossible to describe using Mypy's type language (if it could be specified, it'd be some derivate of a `TypedDict` with higher-kinded dependent types requiring input from the application — again, in a format which currently doesn't exist — and also there's some serialization boundaries which are presently implicit).  However, it needs to be an `Interface` to satisfy the fact that it fits into a `Componentized` and `Componentized` *requires* an `IInterface` as a key.

Discarding some of the jargon here it's like the problem with functions that modify a function signature by adding or removing an expected positional or keyword argument.  There's just no way to tell Mypy "oh, it's like whatever function you got, but wiht these modifications", much as you can't say "well it's like a `Dict` but one which has been explicitly set up to have some keys based on this runtime data".

**Comment by glyph** [comment]
All of that could be more clearly described than "marker interface" so maybe I'll give that a shot as well.

**Comment by glyph** [comment]
`IForm` and `IProtoForm` by contrast should probably be more or less wholesale copies of the public attributes on `ProtoForm` and `Form` since those are "markers" in that the system requires an `Interface` but they are *not* publicly exported; no `IForm` or `IProtoForm` in `interfaces.py`, you'll note.

**Comment by wsanchez** [comment]
> I agree with this, GVOS applies

Garden Valley Oil Services?

**Comment by glyph** [comment]
> > I agree with this, GVOS applies
> 
> 
> 
> Garden Valley Oil Services?

Gross Violation Of Specifications, from the twisted compat policy

**Pull request merged**